infrastructure/tari_script/src/script.rs

```diff
@@ -616,18 +617,24 @@ impl TariScript {
             })
             .collect::<Result<Vec<RistrettoSchnorr>, ScriptError>>()?;
 
-        let mut key_signed = vec![false; public_keys.len()];
         // keep a hashset of unique signatures used to prevent someone putting the same signature in more than once.
         #[allow(clippy::mutable_key_type)]
         let mut sig_set = HashSet::new();
 
         let mut agg_pub_key = RistrettoPublicKey::default();
-        // Check every signature against each public key looking for a valid signature
+
+        // Create an iterator that allows each pubkey to only be checked a single time as they are
+        // removed from the collection when referenced
+        let mut pub_keys = public_keys.iter();
+
+        // Signatures and public keys must be ordered
         for s in &signatures {
-            for (i, pk) in public_keys.iter().enumerate() {
-                if !sig_set.contains(s) && !key_signed[i] && s.verify_raw_canonical(pk, &message) {
-                    // This prevents Alice creating 2 different sigs against her public key
-                    key_signed[i] = true;
+            if pub_keys.len() == 0 {
+                return Ok(None);
+            }
+
+            for pk in pub_keys.by_ref() {
+                if !sig_set.contains(s) && s.verify_raw_canonical(pk, &message) {
```

Comment:
I think you can move `!sig_set.contains(s)` outside of (and immediately prior to) the inner `for` loop to save on hash lookups. This means there is a single check for signature duplication. If the signature is verified and it is added to the hash set, the `break` will short-circuit subsequent lookups anyway in favor of the single check after the loop.

Reply:
Tested this change locally, and all tests pass.